### PR TITLE
Add full support for specifying a version when creating RDS instances

### PR DESCRIPTION
**Pull request by ccostino**
Addresses issues #146 and #147

This changeset adds new functionality to fully support creating and modifying RDS instances with specific versions for PostgreSQL and MySQL instances.  A user can specify a custom parameter to provide a version and as long as it is a supported version in AWS GovCloud (and on a valid upgrade path in the case of modification), that version will be used.

## TODO:

- [x] Fix broker IAM permissions to allow for `rds:DescribeDBEngineVersions` ([found in`cg-provision`](https://github.com/cloud-gov/cg-provision/blob/master/terraform/modules/iam_role_policy/aws_broker/policy.json#L6-L20)) - see https://github.com/cloud-gov/cg-provision/pull/808, https://github.com/cloud-gov/cg-provision/pull/809, and https://github.com/cloud-gov/cg-provision/pull/810
- [x] Test pipeline changes and new smoke tests development
- [x] Make remaining pipeline changes for staging and production
- [x] Test functionality live in development environment

## Changes proposed in this pull request:

- Adds support to create new MySQL instances with a specific version
- Adds minimum and maximum supported version numbers to the plan catalog
- Fixes an issue with the BackupRetentionPeriod catalog property; fixes associated plan information to account for this
- New tests added to cover this functionality
- New smoke tests to cover the full breadth of version support
- Pipeline cleanup
- Removes the specific version for MySQL plans
- Tweaks the storage parameter checks in `update-service` calls to not check for setting the same amount of storage (AWS modify call will ignore the change)
- Fixes an issue with the creation of custom parameter groups when a new instance isn't given a version and is relying on the AWS default
- Swaps custom version parameter to a string to make parsing and checks more consistent

## Security considerations:

- This will allow users to specify versions that may not be current, however they will still be supported in AWS GovCloud as we specify a minimum and maximum
- Removes the specific version defined for MySQL instances so that they now leverage AWS' default

### CODE REVIEW

**Comment by ccostino** [comment]
@cloud-gov/platform-ops, one thought I had here was instead of specifying a min and max version for each RDS plan we offer, we could make a list of approved versions instead for each one.  I think this would make it a bit easier to control and validate that an approved and supported version is chosen, leaving less to chance.  The AWS upgrade path restrictions will still cause some headaches as it is (e.g., you can't upgrade from PostgreSQL 9.6.x to 12.x directly), but this may help alleviate some of that.

We _could_ add more error handling to capture such responses from the AWS API, but I worry about that becoming overly complicated with tons of edge cases trying to be accounted for, and feel that situation is better handled by an operator at that point to better assist a customer.

**Comment by ccostino** [comment]
@soutenniza and I talked through this some more and came to the conclusion that we'd lose some flexibility here if we do that.  Even though the checks on our side would help keep things simple by specifying a major version, it would also mean that a user specifying something like `11.5` for PostgreSQL wouldn't be able to select that version and we'd be back to doing min and max version checks like we are now anyway.  The current implementation keeps things simpler in this regard.

**Comment by ccostino** [comment]
There is currently a bug with this where the `cf update-service` command is not actually passing in the new version number.

**Comment by ccostino** [comment]
It helps when you actually send the right request with the updated parameters... d'oh!

**Comment by ccostino** [comment]
Two issues remain:

- Handling custom parameter groups when updating (currently breaks for MySQL)
- Accounting for the delay in the update operation actually starting vs. what's reporting back when the request is first made

For the first issue, I'm not sure if it's possible to repurpose the existing parameter group for the newer version.  If it is, that could make this easy, but if it's not then we'd likely have to create a new one based on the new engine version, copy over any adjustments made in the existing one, and then delete the existing one when finished.  Since there will likely be a name conflict, I'm thinking it might be best to include the DB version in the custom parameter group name going forward (and using the actual instance ID in the description, too; not the formatted name which strips out the dashes and is a bit confusing).

For the second issue, I'm not entirely sure how to approach that yet.  The update call thinks the request succeeded completely, so it reports it as a success, which prematurely sets the status as `update succeeded` in the broker.

**Comment by ccostino** [comment]
It turns out these last two issues are a bigger thorn to deal with than I originally expected, and after talking this through with @soutenniza some this morning I've of the mind that we should *not* provide the ability for folks to perform major version upgrades on existing instances directly.

My recommendation is to not allow for this, remove the functionality in this PR, and write up documentation that asks the customer to reach out to us via support to work through any major version upgrades directly.

TL;DR version of why:  there's a whole host of steps to go through, especially with MySQL, that make this a non-trivial process:

- [RDS PostgreSQL upgrade instructions](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.PostgreSQL.html#USER_UpgradeDBInstance.PostgreSQL.MajorVersion.Process)
- [RDS MySQL upgrade instructions](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_UpgradeDBInstance.MySQL.html#USER_UpgradeDBInstance.MySQL.Major)

There's not an easy way to automate this without a lot of extra work put into this.  Furthermore, any custom parameter groups that have to be taken care of are an additional wrinkle.

### More details

When I initially researched more of the issue with the custom parameter groups, it seemed like it was solvable.  You can [copy custom parameter groups](https://docs.aws.amazon.com/AmazonRDS/latest/UserGuide/USER_WorkingWithParamGroups.html#USER_WorkingWithParamGroups.Copying), so it seemed to me this flow would be possible in the case of version upgrades:

- Check for the existence of a custom parameter group for the instance; if one exists, check to see if its version family still matches the version of the instance.  In the case of an upgrade request the versions wouldn't match, so an updated custom parameter group is needed
- Copy the existing custom parameter group into a new one that is based off of the correct DB family version
- Remove the old custom parameter group when the upgrade is complete

There are a couple of gotchas with this approach already, though:

- AWS recommends waiting at least 5 minutes for the creation of a new custom parameter group; we could handle this, but it'd be cumbersome
- Not all parameters are copied, and it's not easily determined which parameters are incompatible between versions; we'd run the risk of things not working completely

After digging into this more and reviewing the full upgrade guides further, combined with the timing issues we're seeing in making API calls and the action actually starting, my mind has changed on allowing for this functionality in the broker.

It's enough that we now allow folks to create new instances with a specific version.  We can then support them by either creating a new instance from a backup or updating in place after the customer verifies they're ready, if they're looking to do a major version upgrade.

File: services/rds/broker.go [hunk 1]
```diff
@@ -106,6 +99,17 @@ func (broker *rdsBroker) CreateInstance(c *catalog.Catalog, id string, createReq
 		return planErr
 	}
 
+	// Check to see if there is a version change and if so, check to make sure it's a valid change.
+	if options.Version != "" {
+		// Check to make sure that the version specified is allowed by the plan.
+		if options.Version < plan.MinVersion || options.Version > plan.MaxVersion {
```
Comment (bengerman13) [inline services/rds/broker.go:105]: since we're supporting dotted-releases for mysql, should we have an array of valid versions instead of a min/max?

Comment (bengerman13) [reply services/rds/broker.go:105]: actually, Is this check going to work how we expect now that `Version` is a string?
Golang does a lexical comparison for `<`, so , for instance:

```
8 0 > 8.0 :  false
8.0.1 > 8.0 :  true
8,0 > 8.0 :  false
```

**Review by bengerman13 — DISMISSED** [review]
I think the input validation for versions needs some work now that they're strings, but it's probably good enough for now, if we start working on that as our next task :)

File: services/rds/broker.go (hunk 1 above)
Comment (ccostino) [reply services/rds/broker.go:105]: I had gone back and forth on this a bit too, and read a variety of different things about how Go would process this.

We have to allow for strings because the major releases for MySQL are still with dot notation (e.g., `5.6`,`5.7`, `8.0`), similar to the older PostgreSQL releases (e.g., `9.5`, 9.6`).  In the context of the broker though, we only care about the major version, because AWS will then select a recent minor/patch level version for us.

Given that, and what you've pointed out, I think just setting an array of valid major versions that we will allow for is the better path, and just checking that the value supplied is in that.  If it's not, we can give a more specific error message as well to the user.

I'll get this worked in, I'd rather have it done right before sending it off to the wild! :-)

**Comment by ccostino** [comment]
@bengerman13 and @soutenniza, I've modified how the version checks work and set the plans to now optionally include a list of approved major versions.  This is ready for review again!

**Review by bengerman13 — APPROVED** [review]
